src/Amplitude.php

```diff
@@ -474,49 +396,167 @@ public function getOptOut()
+    }
+
     /**
      * Send the event currently set in $this->event to amplitude
      *
      * Requres $this->event and $this->apiKey to be set, otherwise it throws an exception.
      *
      * @return void
-     * @throws \InternalErrorException If event or api key not set
+     * @throws AmplitudeException If event or api key not set
      */
-    protected function sendEvent()
+    protected function sendEvent(Event $event)
     {
-        if (empty($this->event) || empty($this->apiKey)) {
-            throw new \InternalErrorException('Event or api key not set, cannot send event');
-        }
-        $ch = curl_init(static::AMPLITUDE_API_URL);
-        if (!$ch) {
-            // Could be a number of PHP environment problems, log a critical error
-            $this->logger->critical(
-                'Call to curl_init(' . static::AMPLITUDE_API_URL . ') failed, unable to send Amplitude event'
-            );
+        if ($this->optOut) {
             return;
         }
+
+        if (!empty($this->userId)) {
+            $event->userId = $this->userId;
+        }
+
+        if (!empty($this->deviceId)) {
+            $event->deviceId = $this->deviceId;
+        }
+
+        $this->checkForApiKey();
+        $this->checkForUserIdOrDeviceId($event);
+        $ch = curl_init(static::AMPLITUDE_API_URL);
+        $this->checkCurlHandle($ch);
         $postFields = [
             'api_key' => $this->apiKey,
-            'event' => json_encode($this->event),
+            'event' => json_encode($event),
         ];
         curl_setopt($ch, \CURLOPT_POSTFIELDS, $postFields);
         // Always return instead of outputting response!
         curl_setopt($ch, \CURLOPT_RETURNTRANSFER, true);
         $response = curl_exec($ch);
-        $curlErrno = curl_errno($ch);
-        if ($curlErrno) {
-            $this->logger->critical(
-                'Curl error: ' . curl_error($ch),
-                compact('curlErrno', 'response', 'postFields')
-            );
-        } else {
-            $httpCode = curl_getinfo($ch, \CURLINFO_HTTP_CODE);
-            $this->logger->log(
-                $httpCode === 200 ? Log\LogLevel::INFO : Log\LogLevel::ERROR,
-                'Amplitude HTTP API response: ' . $response,
-                compact('httpCode', 'response', 'postFields')
-            );
-        }
+        $this->checkForCurlError($ch);
+
+        $httpCode = curl_getinfo($ch, \CURLINFO_HTTP_CODE);
+        $this->logger->log(
+            Log\LogLevel::INFO,
+            'Amplitude HTTP API response: ' . $response,
+            compact('httpCode', 'response', 'postFields')
+        );
         curl_close($ch);
     }
+
+    protected function checkForApiKey(): void
+    {
+        if (empty($this->apiKey)) {
+            throw new \LogicException(static::EXCEPTION_MSG_NO_API_KEY);
+        }
+    }
+
+    /**
+     * @param resource $ch
+     * @throws \Exception
+     */
+    protected function checkCurlHandle($ch): void
```

Comment:
If you use return declarations, need to do it consistently, either on all methods are none.  No harm in adding them to all the methods since it will already require a major version bump with all the non-BC changes.